private/bufpkg/bufmodule/module.go

```diff
@@ -89,6 +89,10 @@ func newModuleForBucket(
 	if err != nil {
 		return nil, err
 	}
+	// if the module config has an identity, set the module identity
+	if moduleConfig.ModuleIdentity != nil {
+		options = append(options, ModuleWithModuleIdentity(moduleConfig.ModuleIdentity))
+	}
```

Comment:
This will override the `ModuleWithModuleIdentity[AndCommit]` option provided by the caller of `NewModuleForBucket`, if one was provided:

```go
// NewModuleForBucket returns a new Module. It attempts reads dependencies
// from a lock file in the read bucket.
func NewModuleForBucket(
	ctx context.Context,
	readBucket storage.ReadBucket,
	options ...ModuleOption,
) (Module, error) {
	return newModuleForBucket(ctx, readBucket, options...)
}
```

In practice, this might be OK since the `ModuleIdentity` found in the bucket _should_ always be equivalent to the option provided (i.e. the option is set with the module pin in the module cacher).

However, this is prone to error - if there's ever a reason that the caller of `NewModuleForBucket` would need to _override_ the `ModuleIdentity` that would otherwise be determined by the `buf.yaml` (this condition), then the caller's option will actually be overwritten with what's found in the bucket.

--

One solution out of this is to _prepend_ this option, which allows the other call-site provided option to take effect _after_ this one, hence taking precedence. This is also nuanced behavior though and feels kinda hacky, so we would need to aggressively include comments here if that's our only option. For example,

```go
if moduleConfig.ModuleIdentity != nil {
	// Explain why we're doing this.
	options = append([]ModuleOption{ModuleWithModuleIdentity(moduleConfig.ModuleIdentity)}, options...)
}
```

Alternatively, we could add an argument to the `newModule` constructor below so that it takes a positional `bufmodule.ModuleIdentity` parameter (which may be `nil`). This would be the default value, but it could be naturally overwritten by the option.

```go
// this should only be called by other newModule constructors
func newModule(
	ctx context.Context,
	// must only contain .proto files
	sourceReadBucket storage.ReadBucket,
	moduleIdentity ModuleIdentity,
	dependencyModulePins []bufmoduleref.ModulePin,
	documentation string,
	breakingConfig *bufbreakingconfig.Config,
	lintConfig *buflintconfig.Config,
	options ...ModuleOption,
) (_ *module, retErr error) {
	if err := bufmoduleref.ValidateModulePinsUniqueByIdentity(dependencyModulePins); err != nil {
		return nil, err
	}
	// we rely on this being sorted here
	bufmoduleref.SortModulePins(dependencyModulePins)
	module := &module{
		sourceReadBucket:     sourceReadBucket,
		moduleIdentity:       moduleIdentity,
		dependencyModulePins: dependencyModulePins,
		documentation:        documentation,
		breakingConfig:       breakingConfig,
		lintConfig:           lintConfig,
	}
	for _, option := range options {
		option(module)
	}
	return module, nil
}
```

Both aren't ideal, but I think I'd lean towards the latter option (updating the `newModule` signature) since it's a lot easier to follow. It will require passing in a `nil` argument in the `newModule` referenced in `newModuleForProto` which is a code smell, but we don't have the `ModuleName` in the Protobuf representation of `Module` - it's explicitly passed in via the [ModuleReader](https://github.com/bufbuild/buf/blob/0786cea8690ed81cdd1f5b31b9b5c0529ffdab74/private/bufpkg/bufapimodule/module_reader.go#L67). If we go this route, we should add a comment in-line explaining this at that call-site.

What are your thoughts?

Reply:
After some offline discussion (and some flipflopping, lol), the option we're going to go with is the second, adding a new parameter to `newModule`.

The crux of this discussion is around the intended behaviour for the build. The original intention was that any `ModuleIdentity` set by the call site _should not_ deviate from the `ModuleIdentity` from the config getter. And in practice, this is the case, since there could be a lot of side effects and breakages caused by overriding with a malformed module identity (including to the digest, etc.). That being said, there could be some use cases, such as testing, etc. where the module identity will need to be overridden, and thus we should always respect the value provided by the call site.

Thus, by adding a new parameter, we can explicitly set the module identity when building via the bucket (and keep it `nil` via other build paths), and then still respect the option override.